server/report.go

```diff
@@ -0,0 +1,397 @@
+	b.WriteString(fmt.Sprintf("  %s Down: **%d**", smallRedTriangleDownEmoji, count.Data.Count.Down))
+	b.WriteString(fmt.Sprintf("  %s Unreachable: **%d**", mailboxWithNoMailEmoji, count.Data.Count.Unreachable))
+	b.WriteString(fmt.Sprintf("  %s Pending: **%d**", hourglassFlowingSandEmoji, count.Data.Count.Pending))
+
+	return b.String()
+}
+
+type extractedHost struct {
+	name, state string
+}
+
+type extractedHosts []extractedHost
+
+func (e extractedHosts) Len() int {
+	return len(e)
+}
+
+func (e extractedHosts) Less(i, j int) bool {
+	return e[i].name < e[j].name
+}
+
+func (e extractedHosts) Swap(i, j int) {
+	e[i], e[j] = e[j], e[i]
+}
+
+// extractHosts returns extractedHosts, extracted from hostListData. It returns
+// unknownState state for every host it failed to extract the state.
+func extractHosts(hostListData nagios.HostListData) extractedHosts {
+	var hosts extractedHosts
+
+	for k, v := range hostListData.HostList {
+		var state string
+
+		if err := json.Unmarshal(v, &state); err != nil {
+			state = unknownState
+		}
+
+		hosts = append(hosts, extractedHost{name: k, state: state})
+	}
+
+	return hosts
+}
+
+func formatHostList(list nagios.HostList) string {
+	if list.Result.TypeText != resultTypeTextSuccess {
+		return gettingReportUnsuccessfulMessage("host list", list.Result.TypeText)
+	}
+
+	hosts := extractHosts(list.Data)
+
+	sort.Sort(hosts)
+
+	var b strings.Builder
+
+	b.WriteString("##### HOST LIST\n\n")
+
+	var abnormalOnly bool
+
+	if len(hosts) > maximumReportLength {
+		abnormalOnly = true
+
+		b.WriteString("**Too many hosts. Showing only abnormal state hosts.**\n\n")
+	}
+
+	var linesWritten int
+
+	for _, h := range hosts {
+		if linesWritten == maximumReportLength {
+			b.WriteString("\n\n**Skipped the rest of the hosts.**")
+			return b.String()
+		}
+
+		if abnormalOnly && h.state == upState {
+			continue
+		}
+
+		if linesWritten > 0 {
+			b.WriteRune('\n')
+		}
+
+		b.WriteString(fmt.Sprintf("%s `%s` %s", emoji(h.state), h.name, strings.ToUpper(h.state)))
+		linesWritten++
+	}
+
+	if linesWritten == 0 {
+		b.WriteString("No hosts to show.")
+	}
+
+	return b.String()
+}
+
+func formatServiceCount(count nagios.ServiceCount) string {
+	if count.Result.TypeText != resultTypeTextSuccess {
+		return gettingReportUnsuccessfulMessage("service summary", count.Result.TypeText)
+	}
+
+	var b strings.Builder
+
+	b.WriteString("##### SERVICE SUMMARY\n\n")
+	b.WriteString(fmt.Sprintf("%s OK: **%d**", whiteCheckMarkEmoji, count.Data.Count.Ok))
+	b.WriteString(fmt.Sprintf("  %s Warning: **%d**", warningEmoji, count.Data.Count.Warning))
+	b.WriteString(fmt.Sprintf("  %s Critical: **%d**", bangBangEmoji, count.Data.Count.Critical))
+	b.WriteString(fmt.Sprintf("  %s Unknown: **%d**", questionEmoji, count.Data.Count.Unknown))
+	b.WriteString(fmt.Sprintf("  %s Pending: **%d**", hourglassFlowingSandEmoji, count.Data.Count.Pending))
+
+	return b.String()
+}
+
+type extractedService struct {
```

Comment:
`extractedService` and `extractedHost` are the same struct why do we need both?

Reply:
Thanks for noticing that. These structs have probably differed initially but eventually became the same. I have merged them into one now. :)